src/perf_tool_backend.ml

```diff
@@ -52,20 +59,38 @@ module Recording = struct
       if Perf_capabilities.(do_intersect capabilities configurable_psb_period)
       then
         (* Using Intel Processor Trace with the highest possible granularity. *)
-        "--event=intel_pt/cyc=1,cyc_thresh=1,mtc_period=0/u"
+        [%string
+          "--event=intel_pt/cyc=1,cyc_thresh=1,mtc_period=0/%{perf_selector_of_trace_mode \
+           trace_mode}"]
       else (
         Core.eprintf
           "[Warning: This machine has an older generation processor, timing granularity \
            will be ~1us instead of ~10ns. Consider using a newer machine.]\n\
            %!";
-        "--event=intel_pt//u")
+        [%string "--event=intel_pt//%{perf_selector_of_trace_mode trace_mode}"])
+    in
+    let kcore_opts =
+      match trace_mode, Perf_capabilities.(do_intersect capabilities kcore) with
+      | Trace_userspace, _ -> []
+      | (Trace_include_kernel | Trace_kernel_only), true -> [ "--kcore" ]
+      | (Trace_include_kernel | Trace_kernel_only), false ->
+        (* Strictly speaking, we could recreate tools/perf/perf-with-kcore.sh
+           here instead of bailing. But that's tricky, and upgrading to a newer
+           perf is easier. *)
+        Core.eprintf
+          "[Warning: Installed perf version is too old to support saving a kcore, but \
```

Comment:
This error message should be more specific about what the user actually loses from not having kcore. Is there something more specific you could say here other than "trace may fail to decode"?

Reply:
I don't really know, without getting too in the weeds. In practice what I've seen without the `kcore` is perf emitting erroneous `tr strt` events, which each open a new stack frame that never gets closed (because there's never an associated `tr end`). Supposedly it decoding traces through kernel self-modifying code too, but I don't have concrete examples of that. Perf docs generically say "decode errors" are possible.

>   Another problem that will be experienced is decoder errors. They
     can be caused by inability to access the executed image,
       self-modified or JIT-ed code, or the inability to match side-band
       information (such as context switches and mmaps) which results in
       the decoder not knowing what code was executed.

Reply:
Maybe the answer here is then to loosen the wording. I'm thinking something along the lines of:

Perf userspace tools v5.5 contain an important feature, kcore, that make decoding kernel traces more reliable. In our experience, tracing the kernel *mostly* works without this feature, but you may run into problems if you're trying to trace through self-modifying code (the kernel may do this more than you think).

Reply:
Sure, changed the message.